### PR TITLE
Add root mean square error buffer comparison

### PR DESCRIPTION
**Pull request by dj2**
This CL adds a `RMSE_BUFFER` comparator which can be used with two
buffers to verify their values are within a given tolerance using the
root mean square error comparison method.

Issue #600

### CODE REVIEW

**Comment by dj2** [comment]
Thoughts? This needs some tests around `Buffer::CalculateRMSE` and `Buffer::CalculateDiffs` but I think does what we want? It should also be extensible as we can add new comparators to the CompareBufferCommand as needed.

**Comment by dj2** [comment]
@afd fyi

@dnovillo can you verify I did the RMSE calculation in `Buffer::CalculateRMSE` correctly?

**Comment by hevrard** [comment]
@paulthomson can you have a look?
@dj2 Paul has been working on this in GraphicsFuzz, and he has more bandwidth than me for this now.

File: src/buffer.cc [hunk 1]
```diff
@@ -53,6 +54,41 @@ T* ValuesAs(uint8_t* values) {
...
+    return Sub<uint8_t>(buf1, buf2);
+  if (comp.IsUint16())
+    return Sub<uint16_t>(buf1, buf2);
+  if (comp.IsUint32())
+    return Sub<uint32_t>(buf1, buf2);
+  if (comp.IsUint64())
+    return Sub<uint64_t>(buf1, buf2);
+  // TOOD(dsinclair): Handle float16 ...
+  if (comp.IsFloat16())
+    return 0.f;
```
Comment (paulthomson) [inline src/buffer.cc]: Should we assert false here? Or was this supposed to be implemented before merging?

Comment (dj2) [reply src/buffer.cc]: Should probably assert. I don't understand how float16 works at the moment so I'm skipping it for now. We have to verify all the float16 code in amber in general anyway.

Comment (dj2) [reply src/buffer.cc]: done

File: docs/amber_script.md [hunk 2]
```diff
@@ -446,6 +447,10 @@ EXPECT {buffer_name} IDX _x_in_pixels_ _y_in_pixels_ \
 
 # Checks that |buffer_1| contents are equal to those of |buffer_2|
 EXPECT {buffer_1} EQ_BUFFER {buffer_2}
+
+# Checks that the Root Mean Square Error when comparing |buffer_1| to
+# |buffer_2| is less than or equal too |tolerance|
```
Comment (dneto0) [inline docs/amber_script.md]: The syntax looks ok to me.  I'd clarify that |tolerance| is unitless.
E.g. when one buffer is an unbiased estimater of the other, then RMSE is the standard deviation, which is unitless.

File: src/buffer.cc [hunk 3]
```diff
@@ -111,6 +150,58 @@ Result Buffer::IsEqual(Buffer* buffer) const {
...
+    return Result{"Buffers have a different size"};
+  if (buffer->width_ != width_)
+    return Result{"Buffers have a different width"};
+  if (buffer->height_ != height_)
+    return Result{"Buffers have a different height"};
+  if (buffer->ValueCount() != ValueCount())
+    return Result{"Buffers have a different number of values"};
+
+  auto diffs = CalculateDiffs(buffer);
+  float sum = 0.0;
```
Comment (dneto0) [inline src/buffer.cc]: I'd do this in double.
Sums of squares can get pretty big.

We _could_ go even further and normalize.:   E.g. divide  each diff by the _mean_ of the diffs, and then compute RMSE, then remultiply by the absolute value of the mean.
But that's probably excessive for now.

**Pull request merged**